### PR TITLE
Fix PR #139

### PR DESCRIPTION
**Pull request by chaserli**
* Suppressed all capture events for buildings considered vehicles.
* Use `MindControl.Anim` for buildings deployed from mindcontrolled vehicles

### CODE REVIEW

**Comment by Starkku** [comment]
There are a couple of issues with the MC ring transfer fix here.

1. Checking weapon via `SelectWeapon` to figure out the mind control ring animation to use is not reliable. It is possible for the weapon used against specific target to change on the fly (f.ex shields, NoAmmoWeapon, eventually AE) and thus it can give wrong results. It is better to remember the ring animation used prior to deploy and use that.

2. Permanent MC ring anim is not correctly recreated or adjusted and thus is only transferred one way and it will have wrong offset.

Something like the following can be used to ensure correct animation is used, and that permanent MC has the ring animation set correctly (it is recreated here instead of simply transferred over for consistency with regular MC). When deploying from vehicle to building, the vehicle will still have `TechnoClass->MindControlRingAnim` but other way around the pointer has been invalidated because the structure has been limboed, hence why one of the hooks in undeploying/selling function has been adjusted to remember the MC anim (by reusing the mechanism added for recreating MC anims when cloaking/decloaking, which will also allow the correct MC anim to be used in case unit is deployed when cloaked and the ring is not currently present).

```cpp
void TechnoExt::TransferMindControlOnDeploy(TechnoClass* pTechnoFrom, TechnoClass* pTechnoTo)
{
	auto const pAnimType = pTechnoFrom->MindControlRingAnim ?
		pTechnoFrom->MindControlRingAnim->Type : TechnoExt::ExtMap.Find(pTechnoFrom)->MindControlRingAnimType;

	if (auto Controller = pTechnoFrom->MindControlledBy)
	{
		if (auto Manager = Controller->CaptureManager)
		{
			CaptureManagerExt::FreeUnit(Manager, pTechnoFrom, true);

			if (CaptureManagerExt::CaptureUnit(Manager, pTechnoTo, false, pAnimType, true))
			{
				if (auto pBld = abstract_cast<BuildingClass*>(pTechnoTo))
				{
					pBld->BeginMode(BStateType::Construction);
					pBld->QueueMission(Mission::Construction, false);
				}
			}
			else
			{
				int nSound = pTechnoTo->GetTechnoType()->MindClearedSound;
				if (nSound == -1)
					nSound = RulesClass::Instance->MindClearedSound;
				if (nSound != -1)
					VocClass::PlayIndexAtPos(nSound, pTechnoTo->Location);
			}

		}
	}
	else if (auto MCHouse = pTechnoFrom->MindControlledByHouse)
	{
		pTechnoTo->MindControlledByHouse = MCHouse;
		pTechnoFrom->MindControlledByHouse = nullptr;
	}
	else if (pTechnoFrom->MindControlledByAUnit) // Perma MC
	{
		pTechnoTo->MindControlledByAUnit = true;

		if (pTechnoTo->MindControlRingAnim)  // This check might not be necessary as there shouldn't be a prior anim on the techno.
			pTechnoTo->MindControlRingAnim->TimeToDie = 1;

		auto const pBuilding = abstract_cast<BuildingClass*>(pTechnoTo);
		CoordStruct location = pTechnoTo->GetCoords();

		if (pBuilding)
			location.Z += pBuilding->Type->Height * Unsorted::LevelHeight;
		else
			location.Z += pTechnoTo->GetTechnoType()->MindControlRingOffset;

		if (auto const pAnim = GameCreate<AnimClass>(pAnimType, location, 0, 1))
		{
			pTechnoTo->MindControlRingAnim = pAnim;

			if (pBuilding)
				pAnim->ZAdjust = -1024;

			pAnim->SetOwnerObject(pTechnoTo);
		}

	}
}

DEFINE_HOOK(0x449E2E, BuildingClass_Mi_Selling_CreateUnit, 0x6)
{
	GET(BuildingClass*, pStructure, EBP);
	R->ECX<HouseClass*>(pStructure->GetOriginalOwner());

	// Remember MC ring animation.
	if (pStructure->IsMindControlled())
	{
		auto const pTechnoExt = TechnoExt::ExtMap.Find(pStructure);
		pTechnoExt->UpdateMindControlAnim();
	}

	return 0x449E34;
}
```

**Review by Starkku — CHANGES_REQUESTED** [review]
There are a couple of issues with the MC ring transfer fix here.

1. Checking weapon via `SelectWeapon` to figure out the mind control ring animation to use is not reliable. It is possible for the weapon used against specific target to change on the fly (f.ex shields, NoAmmoWeapon, eventually AE) and thus it can give wrong results. It is better to remember the ring animation used prior to deploy and use that.

2. Permanent MC ring anim is not correctly recreated or adjusted, also the MC status was not actually properly transferred.

Something like the following can be used to ensure correct animation is used, and that permanent MC has the ring animation set correctly (it is recreated here instead of simply transferred over for consistency with regular MC). When deploying from vehicle to building, the vehicle will still have `TechnoClass->MindControlRingAnim` but other way around the pointer has been invalidated because the structure has been limboed, hence why one of the hooks in undeploying/selling function has been adjusted to remember the MC anim (by reusing the mechanism added for recreating MC anims when cloaking/decloaking, which will also allow the correct MC anim to be used in case unit is deployed when cloaked and the ring is not currently present).

```cpp
void TechnoExt::TransferMindControlOnDeploy(TechnoClass* pTechnoFrom, TechnoClass* pTechnoTo)
{
	auto const pAnimType = pTechnoFrom->MindControlRingAnim ?
		pTechnoFrom->MindControlRingAnim->Type : TechnoExt::ExtMap.Find(pTechnoFrom)->MindControlRingAnimType;

	if (auto Controller = pTechnoFrom->MindControlledBy)
	{
		if (auto Manager = Controller->CaptureManager)
		{
			CaptureManagerExt::FreeUnit(Manager, pTechnoFrom, true);

			if (CaptureManagerExt::CaptureUnit(Manager, pTechnoTo, false, pAnimType, true))
			{
				if (auto pBld = abstract_cast<BuildingClass*>(pTechnoTo))
				{
					pBld->BeginMode(BStateType::Construction);
					pBld->QueueMission(Mission::Construction, false);
				}
			}
			else
			{
				int nSound = pTechnoTo->GetTechnoType()->MindClearedSound;
				if (nSound == -1)
					nSound = RulesClass::Instance->MindClearedSound;
				if (nSound != -1)
					VocClass::PlayIndexAtPos(nSound, pTechnoTo->Location);
			}

		}
	}
	else if (auto MCHouse = pTechnoFrom->MindControlledByHouse)
	{
		pTechnoTo->MindControlledByHouse = MCHouse;
		pTechnoFrom->MindControlledByHouse = nullptr;
	}
	else if (pTechnoFrom->MindControlledByAUnit) // Perma MC
	{
		pTechnoTo->MindControlledByAUnit = true;

		if (pTechnoTo->MindControlRingAnim)  // This check might not be necessary as there shouldn't be a prior anim on the techno.
			pTechnoTo->MindControlRingAnim->TimeToDie = 1;

		auto const pBuilding = abstract_cast<BuildingClass*>(pTechnoTo);
		CoordStruct location = pTechnoTo->GetCoords();

		if (pBuilding)
			location.Z += pBuilding->Type->Height * Unsorted::LevelHeight;
		else
			location.Z += pTechnoTo->GetTechnoType()->MindControlRingOffset;

		if (auto const pAnim = GameCreate<AnimClass>(pAnimType, location, 0, 1))
		{
			pTechnoTo->MindControlRingAnim = pAnim;

			if (pBuilding)
				pAnim->ZAdjust = -1024;

			pAnim->SetOwnerObject(pTechnoTo);
		}

	}
}

DEFINE_HOOK(0x449E2E, BuildingClass_Mi_Selling_CreateUnit, 0x6)
{
	GET(BuildingClass*, pStructure, EBP);
	R->ECX<HouseClass*>(pStructure->GetOriginalOwner());

	// Remember MC ring animation.
	if (pStructure->IsMindControlled())
	{
		auto const pTechnoExt = TechnoExt::ExtMap.Find(pStructure);
		pTechnoExt->UpdateMindControlAnim();
	}

	return 0x449E34;
}
```

**Review by Starkku — APPROVED** [review]

**Review by Starkku — APPROVED** [review]
Took a liberty of implementing the feedback from my prior review and attempted to address the problem with building anims playing prematurely etc. after deploying a mind controlled vehicle. Unsure if the fix for that is complete but it appears to do the job.

File: src/Ext/CaptureManager/Body.h [hunk 1]
```diff
@@ -15,7 +15,7 @@ class CaptureManagerExt
 	static bool CanCapture(CaptureManagerClass* pManager, TechnoClass* pTarget);
 	static bool FreeUnit(CaptureManagerClass* pManager, TechnoClass* pTarget, bool bSilent = false);
 	static bool CaptureUnit(CaptureManagerClass* pManager, TechnoClass* pTarget,
-		bool bRemoveFirst, AnimTypeClass* pControlledAnimType = RulesClass::Instance->ControlledAnimationType);
+		bool bRemoveFirst, AnimTypeClass* pControlledAnimType = RulesClass::Instance->ControlledAnimationType, bool bSilent = false);
```
Comment (MortonPL) [inline src/Ext/CaptureManager/Body.h:18]:
```suggestion
		bool bRemoveFirst, AnimTypeClass* pControlledAnimType = RulesClass::Instance->ControlledAnimationType, bool silent = false);
```

Aside from p for pointers, I don't think we use (and should use) the Hungarian notation for variables/arguments.

File: src/Ext/CaptureManager/Body.h [hunk 2]
```diff
@@ -15,7 +15,7 @@ class CaptureManagerExt
 	static bool CanCapture(CaptureManagerClass* pManager, TechnoClass* pTarget);
 	static bool FreeUnit(CaptureManagerClass* pManager, TechnoClass* pTarget, bool bSilent = false);
```
Comment (MortonPL) [inline src/Ext/CaptureManager/Body.h:16]:
```suggestion
	static bool FreeUnit(CaptureManagerClass* pManager, TechnoClass* pTarget, bool silent = false);
```

File: src/Ext/CaptureManager/Body.cpp [hunk 3]
```diff
@@ -90,7 +90,7 @@ bool CaptureManagerExt::CaptureUnit(CaptureManagerClass* pManager, TechnoClass*
 			pManager->ControlNodes.AddItem(pControlNode);
 			pControlNode->LinkDrawTimer.Start(RulesClass::Instance->MindControlAttackLineFrames);
 
-			if (pTarget->SetOwningHouse(pManager->Owner->Owner))
+			if (pTarget->SetOwningHouse(pManager->Owner->Owner,!bSilent))
```
Comment (MortonPL) [inline src/Ext/CaptureManager/Body.cpp:93]:
```suggestion
			if (pTarget->SetOwningHouse(pManager->Owner->Owner, !silent))
```

File: src/Ext/CaptureManager/Body.cpp [hunk 4]
```diff
@@ -65,7 +65,7 @@ bool CaptureManagerExt::FreeUnit(CaptureManagerClass* pManager, TechnoClass* pTa
 }
 
 bool CaptureManagerExt::CaptureUnit(CaptureManagerClass* pManager, TechnoClass* pTarget,
-	bool bRemoveFirst, AnimTypeClass* pControlledAnimType)
+	bool bRemoveFirst, AnimTypeClass* pControlledAnimType, bool bSilent)
```
Comment (MortonPL) [inline src/Ext/CaptureManager/Body.cpp:68]:
```suggestion
	bool bRemoveFirst, AnimTypeClass* pControlledAnimType, bool silent)
```

File: src/Ext/CaptureManager/Body.cpp [hunk 5]
```diff
@@ -47,7 +47,7 @@ bool CaptureManagerExt::FreeUnit(CaptureManagerClass* pManager, TechnoClass* pTa
 				auto pOriginOwner = pNode->OriginalOwner->Defeated ?
 					HouseClass::FindNeutral() : pNode->OriginalOwner;
 
-				pTarget->SetOwningHouse(pOriginOwner);
+				pTarget->SetOwningHouse(pOriginOwner, !bSilent);
```
Comment (MortonPL) [inline src/Ext/CaptureManager/Body.cpp:50]:
```suggestion
				pTarget->SetOwningHouse(pOriginOwner, !silent);
```

There are more uses of `bSilent` above this diff, but I can't mark them in suggestions, so they have to be changed manually.

**Review by MortonPL — COMMENTED** [review]
A minor comment about names - we don't use the Hungarian notation.

**Pull request merged**